helper/schema/schema.go

```diff
@@ -1431,6 +1480,79 @@ func (m schemaMap) validateConflictingAttributes(
 	return nil
 }
 
+func removeDuplicates(elements []string) []string {
+	encountered := make(map[string]struct{}, 0)
+	result := []string{}
+
+	for v := range elements {
+		if _, ok := encountered[elements[v]]; !ok {
+			encountered[elements[v]] = struct{}{}
+			result = append(result, elements[v])
+		}
+	}
+
+	return result
+}
+
+func validateExactlyOneAttributes(
+	k string,
+	schema *Schema,
+	c *terraform.ResourceConfig) error {
+
+	if len(schema.ExactlyOneOf) == 0 {
+		return nil
+	}
+
+	allKeys := removeDuplicates(append(schema.ExactlyOneOf, k))
+	sort.Strings(allKeys)
+	specified := make([]string, 0)
+	for _, exactlyOneOfKey := range allKeys {
+		if raw, ok := c.Get(exactlyOneOfKey); ok {
+			if raw == hcl2shim.UnknownVariableValue {
+				// An unknown value might become unset (null) once known, so
+				// we must defer validation until it's known.
+				continue
+			}
+			specified = append(specified, exactlyOneOfKey)
+		}
+	}
+
+	if len(specified) == 0 {
+		return fmt.Errorf("%q: one of `%s` must be specified", k, strings.Join(allKeys, ","))
+	}
+
+	if len(specified) > 1 {
+		return fmt.Errorf("%q: only one of `%s` can be specified, but `%s` were specified.", k, strings.Join(allKeys, ","), strings.Join(specified, ","))
+	}
+
+	return nil
+}
+
+func validateAtLeastOneAttributes(
+	k string,
+	schema *Schema,
+	c *terraform.ResourceConfig) error {
+
+	if len(schema.AtLeastOneOf) == 0 {
+		return nil
+	}
+
+	allKeys := removeDuplicates(append(schema.AtLeastOneOf, k))
+	sort.Strings(allKeys)
```

Comment:
same as above do we need to sort?

Reply:
I'm guessing it makes for nicer printout in the error message having the keys being sorted?

Reply:
Exactly!